### ISSUE TITLE
[question] Conan 1.x: How do I avoid passing on a default dependency option?

### ISSUE THREAD

**Issue by RealFrogo** [conan-io/conan#15309]
### What is your question?

Hi!

We are in the process of introducing conan into our build environment (for the time being we are stuck with 1.x, as things move slowly). It was new to me as well, so I'm still trying to find my footing.

Right now I'm facing the following problem (code represents snippets of the respective conanfile.py files), and I'm not sure if it's intended or a bug:

I have a base package pkg.a with boolean option x, which defaults to false:

```
class PkgA(ConanFile):
    name = "pkg.a"
    [...]
    options = {
        "x": [True, False]
    }
    default_options = {
        "x": False
    }
```

I have package pkg.b, which depends on pkg.a with option x=true to be built:

```
class PkgB(ConanFile):
    name = "pkg.b"
    requires = [("pkg.a")]
    [...]
    default_options = {
        "pkg.a:x": True
    }
```

Important for later: the resulting libraries don't need option x to be true any more when being used and linked to pkg.a.

Now I have packages pkg.c, pkg.d and pkg.e, which depend on pkg.a with option x=false as well as pkg.b to be built:

```
class PkgC/D/E(ConanFile):
    name = "pkg.c/d/e"
    requires = [("pkg.a"),
                ("pkg.b")]
    [...]
    default_options = {
        "pkg.a:x": False
    }
```

When I set the dependency of e.g. pkg.c to pkg.a and pkg.b without explicity setting option pkg.a:x to false as shown, it is set to true, even though its default is false (as I learned the hard way). I guess that's because its overwritten by the dependency of pkg.b? It doesn't mention that in the output of conan install, however.

My question is, is that the expected behaviour? Because it doesn't seem obvious to me but rather missleading, and it took me a while to figure out my errors in building pkg.c were due to accidentally linking the wrong version of pkg.a's library. 

However, if it is the expected behaviour, can I avoid always having to explicitly reset pkg.a:x to false, simply by telling pkg.b to not pass its option setting along? The thing is, in our case every package which depends on pkg.b needs pkg.a:x to be false. As mentioned, once built it doesn't matter to pkg.b anymore if the option is set, but it matters to the other packages, but I'd like to avoid having to always think of resetting the option when depending on package pkg.b as well, as this strikes me as an invitation for uneccessary errors, which sometimes can be quite hard to find (as already mentioned).

What I'd like is to tell pkg.b when creating the package to forget about this option which it only needed to be built, so it doesn't get passed along. However, simply resetting it in the package() method didn't work, and so far my research was fruitless, as every example regarding passing along options or inheritating them always only considered two levels, not three as in my case.

One solution which was proposed to me was to skip setting the default option in pkg.b and pass it via CLI instead (or use profiles), but that conflicts with our approach to keep all package related information in the conanfile.py, so we can use general creation calls in our build environment.

Any help would be apreciated!

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @RealFrogo 

Thanks for  your question.

> My question is, is that the expected behaviour? Because it doesn't seem obvious to me but rather missleading, and it took me a while to figure out my errors in building pkg.c were due to accidentally linking the wrong version of pkg.a's library.

Yes, this is expected behavior. There is one design rule in Conan that is that the user or the most downstream consumer always have priority. So whatever options the user define on command line via ``-o pkg*:shared=True`` or in a profile ``[options]`` section and passed with ``-pr=profile`` will have precedence over anything that recipes define. And if the user defines a recipe and in the recipe they write ``default_options = {pkg*:shared: True}`` that also will have precedence.

This follows the principle of "least surprise". The developer/user write those and they totally expect that to be respected. If the dependencies will not get those options values the users will be very frustrated and confused.

> My question is, is that the expected behaviour? Because it doesn't seem obvious to me but rather missleading, and it took me a while to figure out my errors in building pkg.c were due to accidentally linking the wrong version of pkg.a's library.

If you want to protect against this kind of errors, the recommended way is to write some ``validate()`` or ``validate_build()`` in the recipes that checks their values or their dependencies options values, and then raise accordingly with "I cannot work with this option value", forcing the consumers to define the right options for the occasion.

> However, if it is the expected behaviour, can I avoid always having to explicitly reset pkg.a:x to false, simply by telling pkg.b to not pass its option setting along? The thing is, in our case every package which depends on pkg.b needs pkg.a:x to be false. As mentioned, once built it doesn't matter to pkg.b anymore if the option is set, but it matters to the other packages, but I'd like to avoid having to always think of resetting the option when depending on package pkg.b as well, as this strikes me as an invitation for uneccessary errors, which sometimes can be quite hard to find (as already mentioned).

It will be very useful to understand what is the option in real world. What is configuring, how is it being used at build time, specially how it does affect the binary of ``pkgb`` and the consumers of ``pkgb``, why it has to get that value only when building that package and not when consumed, because this seems counterintuitive for the problem that the options are modeling. The most classic example is a package being ``shared=True``. If you want ``pkga`` to be ``shared`` that is always the same, it doesn't make sense to have ``pkga`` shared while building ``pkgb`` but then not shared (static) while being consumed by other packages, because that might create linking problems in the graph.

**Comment by RealFrogo** [comment]
Thank you for the quick reply!

I get why this is expected behaviour and can accept that, thank you for the explanation. I guess I have to live with my surprise being on of the least kind ;)

Anyway, using validate() might be a sensible approach here. So far I disregarded that method, as most of the stuff we're doing is straight forward and there simply didn't seem to be the need for explicitly validating what seemed to be implicitly clear, and having less code seemed to be the way to go so far. Looks like we've been proven wrong.

For your information, the reason why building and using pkg.b differs is because of its structure and that the option in question changes the compiler flags used for pkg.a. Pkg.b is mostly a program running on its own (part of a bigger system of interconnected programs), but which provides a small library to link against for communication with said program (this is legacy architecture and cannot be changed). And those compiler flags are relevant for the program itself (which links to pkg.a), which is not available to depending packages, but built in the same run as the small library, which is, but to which the compiler flag makes no difference (as it only needs the headers from pkg.a, which are the same no matter the option). The library is part of the program as well, which is why they are built together and always have been. My job is "merely" to introduce a decent package management system for all this, which is why I'm trying to figure out how best to approach that with conan, without actually being able to change the programs themselves. Because obviously splitting it up might do the trick, I think, but I'm not sure I'm allowed or even able to do so.

Anyway, thanks for your help, it certainly gave me a push in the right direction. If the last paragraph doesn't give you an additional idea for me to try out (I hope I made more or less clear what the problem is), feel free to close the issue.

**Comment by memsharded** [comment]
That helps, thanks for the feedback!

> For your information, the reason why building and using pkg.b differs is because of its structure and that the option in question changes the compiler flags used for pkg.a. Pkg.b is mostly a program running on its own (part of a bigger system of interconnected programs), but which provides a small library to link against for communication with said program (this is legacy architecture and cannot be changed). And those compiler flags are relevant for the program itself (which links to pkg.a), which is not available to depending packages, but built in the same run as the small library, which is, but to which the compiler flag makes no difference (as it only needs the headers from pkg.a, which are the same no matter the option).

So if I understand it correctly you need 2 binaries of ``pkga``, one binary built with the specific compiler flags that the application in ``pkgb`` needs, and another binary (different ``package_id``) built with the regular flags that all the other packages consuming it will link against? Is this the use case? If that is the case, at the end of the day 2 builds (2 different ``conan create`` are needed for ``pkga``, and if that is the case, then probably the "special" exception of doing this for this package can easily inject in the job command line the extra ``-o pka:....`` option to force the special flags)

I'd like to suggest an extra thing to consider: It seems this could be one of the cases where ``pkgb`` could have a ``self.requires("pkga/...", visible=False, options={"myoption": "value"])`` might make sense. Specially if the small library in ``pkgb`` does not link with ``pkga`` at all, and only the executable does. That means that you don't want to propagate the linkage requirement of ``pkga`` to the consumers of ``pkgb``, and ``visible=False`` will do exactly that, making this requirement "private".  This is important, as it seems your dependency graph should actually have 2 different instances of the ``pkga`` dependency, one built with the option True and the other with the option False. It is not possible to have 2 difference instances of the same package, in the same dependency graph, unless one of them is not visible. So let me try to summarize, this is what I would try:

- ``pkga`` option default is ``False``
- You need to build ``pkga`` with both option=True and option=False, to have those 2 different binaries
- ``pkgb`` will have a private requirement to ``pkga`` via ``self.requires("pkga/...", visible=False, options={`"option": True})``
- The rest of dependencies shouldn't define any option value for ``pkga``, they will use the default which is ``False``. 

I think this should work, please let me know if this makes sense, and if you have any further question. I can help trying to put together a proof of concept via a unit test or something similar if necessary. I prefer to keep the issue open until this is completely and a full solution is working for you. Thanks!

**Comment by RealFrogo** [comment]
Thank you very much, this seems to be exactly what I'm looking for! 

You are right, pkga does have different packages based on the option, but I wasn't aware of the "private" aspect of the requirements anymore (read about it once and forgot about it again, as it wasn't relevant back then...). This might indeed solve this problem in a nice and clean way, I'll give it a try and let you know!

(Won't be today anymore, however, I'm about to call it a day, in my timezone it's 5:30 pm)

**Comment by memsharded** [comment]
Ok, sounds great!

**Comment by RealFrogo** [comment]
Ok, I tinkered some more and finally got a working solution which lets me wonder why I didn't think of it sooner... but let me tell you what I figured out:

Your approach unfortunately didn't work, as neither the "visible" nor the "options" setting was recognized (probably because I still have to use Conan 1.x? Theres no corresponding documentation here either: https://docs.conan.io/1/reference/conanfile/methods.html#requirements) when building pkgb. The former could be compensated by using "private" instead, which led to interesting results (more to it later), but the latter simply was ignored, no matter if I wrote "myOption" or "pkga:myOption" (as you'd do in the default_options). 

What I tried instead was moving pkga:myOption in pkgb's conanfile.py from default_options to the configure() method, but with the same result at first. It got interesting, however, once I tried the "private" setting, because it does get recognized in some way. Now, when calling `conan install` on pkgc (which depends on pkga and pkgb), it references *both* packages of pkga, the one with myOption=True as well as the one with myOption=False. The one used by pkgb (with myOption=True) is shown as skipped, however! Interestingly enough, when using the validate() method to check that myOption is false, it still raises an exception.

So even if did seemed to be a step in the right direction, I wasn't satisifed (just like the validate() method) and didn't bother to check if the compiled program worked anyway. But as using the configure() method did work as such, I now of course had the possibility to use conditionals. No idea why I didn't think of this earlier, but now I can simply use an option of pkgb to make sure that when creating the package for other programs to use (and only then), I simply don't set myOption to True. I planned to introduce such an option anyway to differentiate between packages containing only the lib and the full program, so that is now opviously the right approach, as I see it.

Anyway, long story short, I finally got what I wanted, and learned some more about Conan on the way, which I already know will be of use later on. Thank you for your help and time!

**Comment by memsharded** [comment]
> Your approach unfortunately didn't work, as neither the "visible" nor the "options" setting was recognized (probably because I still have to use Conan 1.x?

Oh, ok, sorry, I was thinking for that in 2.0, visible and options in ``requires()`` is indeed a 2.0 feature. 

> when building pkgb. The former could be compensated by using "private" instead, which led to interesting results (more to it later),

Yes, ``private`` is more or less the equivalent to ``visible=True``, but it is true that we don't encourage it in 1.X, because it has some not great behavior

> So even if did seemed to be a step in the right direction, I wasn't satisifed (just like the validate() method) and didn't bother to check if the compiled program worked anyway. But as using the configure() method did work as such, I now of course had the possibility to use conditionals. No idea why I didn't think of this earlier, but now I can simply use an option of pkgb to make sure that when creating the package for other programs to use (and only then), I simply don't set myOption to True. I planned to introduce such an option anyway to differentiate between packages containing only the lib and the full program, so that is now opviously the right approach, as I see it.

Ok, happy that you found a way to manage this. When moving to Conan 2.0 (recommended to prioritize), as this would be complicated to find a simultaneously jointly compatible 1.X-2.0 syntax, you can handle this with an ugly ``if conan_version >= "2": .... else: ...``

> Anyway, long story short, I finally got what I wanted, and learned some more about Conan on the way, which I already know will be of use later on. Thank you for your help and time!

Very happy to help, I am closing this ticket as responded, but don't hesitate to open new tickets for any further question or issue you might have, thanks!
